delta-app/src/main/java/io/cdap/delta/app/service/AssessmentHandler.java

```diff
@@ -288,6 +285,88 @@ public void getState(HttpServiceRequest request, HttpServiceResponder responder,
+      DeltaWorkerId id = new DeltaWorkerId(new DeltaPipelineId(namespaceName, app, generation), instanceid);
+      OffsetAndSequence offset = getStateStore().readOffset(id);
+      responder.sendString(GSON.toJson(offset));
+    });
+  }
+
+  //Upserts in the delta_offset_store table for offset data.
+  @PUT
+  @Path("v1/contexts/{namespace}/apps/{app}/generations/{generation}/instanceIds/{instanceid}/offset")
+  public void saveOffsetData(HttpServiceRequest request, HttpServiceResponder responder,
+                             @PathParam("namespace") String namespaceName,
+                             @PathParam("app") String app,
+                             @PathParam("generation") long generation,
+                             @PathParam("instanceid") int instanceid) {
+    respond(namespaceName, responder, (namespace) -> {
+      DeltaWorkerId id = new DeltaWorkerId(new DeltaPipelineId(namespaceName, app, generation), instanceid);
+      String body = StandardCharsets.UTF_8.decode(request.getContent()).toString();
+      getStateStore().writeOffset(id, GSON.fromJson(body, OffsetAndSequence.class));
+      responder.sendStatus(HttpURLConnection.HTTP_OK);
+    });
+  }
+
+  //Looks up the delta_state_store table for the given state
+  @GET
+  @Path("v1/contexts/{namespace}/apps/{app}/generations/{generation}/instanceIds/{instanceid}/states/{state}")
+  public void getStateData(HttpServiceRequest request, HttpServiceResponder responder,
+                        @PathParam("namespace") String namespaceName,
+                        @PathParam("app") String app,
+                        @PathParam("generation") long generation,
+                        @PathParam("instanceid") int instanceid,
+                        @PathParam("state") String state) {
+    respond(namespaceName, responder, (namespace) -> {
+      DeltaWorkerId id = new DeltaWorkerId(new DeltaPipelineId(namespaceName, app, generation), instanceid);
+      byte[] data = getStateStore().readState(id, state);
+      responder.sendString(new String(data, StandardCharsets.UTF_8));
+    });
+  }
+
+  // Upserts in the delta_state_store table for state data.
+  @PUT
+  @Path("v1/contexts/{namespace}/apps/{app}/generations/{generation}/instanceIds/{instanceid}/states/{state}")
+  public void saveStateData(HttpServiceRequest request, HttpServiceResponder responder,
+                         @PathParam("namespace") String namespaceName,
+                         @PathParam("app") String app,
+                         @PathParam("generation") long generation,
+                         @PathParam("instanceid") int instanceid,
+                         @PathParam("state") String state) {
+
+    respond(namespaceName, responder, (namespace) -> {
+      DeltaWorkerId id = new DeltaWorkerId(new DeltaPipelineId(namespaceName, app, generation), instanceid);
+      ByteBuffer byteBuffer = request.getContent();
+      byte[] data = new byte[byteBuffer.remaining()];
+      byteBuffer.get(data);
```

Comment:
This way will not reset the ByteBuffer position. Use `Bytes.toBytes()` instead of manually doing this.